src/NHibernate.Test/MultiTenancy/DatabaseStrategyNoDbSpecificFixture.cs

```diff
@@ -0,0 +1,348 @@
+			Assert.That(Sfi.Statistics.SecondLevelCacheHitCount, Is.EqualTo(1));
+		}
+
+		[Test]
+		public void SecondLevelCacheSeparationPerTenant()
+		{
+			using (var sesTen1 = OpenTenantSession("tenant1"))
+			{
+				var entity = sesTen1.Get<Entity>(_id);
+			}
+
+			Sfi.Statistics.Clear();
+			using (var sesTen2 = OpenTenantSession("tenant2"))
+			{
+				var entity = sesTen2.Get<Entity>(_id);
+			}
+
+			Assert.That(Sfi.Statistics.PrepareStatementCount, Is.EqualTo(1));
+			Assert.That(Sfi.Statistics.SecondLevelCacheHitCount, Is.EqualTo(0));
+		}
+
+		[Test]
+		public void QueryCacheReusedForSameTenant()
+		{
+			using (var sesTen1 = OpenTenantSession("tenant1"))
+			{
+				var entity = sesTen1.Query<Entity>().WithOptions(x => x.SetCacheable(true)).Where(e => e.Id == _id).SingleOrDefault();
+			}
+
+			Sfi.Statistics.Clear();
+			using (var sesTen2 = OpenTenantSession("tenant1"))
+			{
+				var entity = sesTen2.Query<Entity>().WithOptions(x => x.SetCacheable(true)).Where(e => e.Id == _id).SingleOrDefault();
+			}
+
+			Assert.That(Sfi.Statistics.PrepareStatementCount, Is.EqualTo(0));
+			Assert.That(Sfi.Statistics.QueryCacheHitCount, Is.EqualTo(1));
+		}
+
+		[Test]
+		public void QueryCacheSeparationPerTenant()
+		{
+			using (var sesTen1 = OpenTenantSession("tenant1"))
+			{
+				var entity = sesTen1.Query<Entity>().WithOptions(x => x.SetCacheable(true)).Where(e => e.Id == _id).SingleOrDefault();
+			}
+
+			Sfi.Statistics.Clear();
+			using (var sesTen2 = OpenTenantSession("tenant2"))
+			{
+				var entity = sesTen2.Query<Entity>().WithOptions(x => x.SetCacheable(true)).Where(e => e.Id == _id).SingleOrDefault();
+			}
+
+			Assert.That(Sfi.Statistics.PrepareStatementCount, Is.EqualTo(1));
+			Assert.That(Sfi.Statistics.QueryCacheHitCount, Is.EqualTo(0));
+		}
+
+		[Test]
+		public void TenantSessionIsSerializableAndCanBeReconnected()
+		{
+			ISession deserializedSession = null; 
+			using (var sesTen1 = OpenTenantSession("tenant1"))
+			{
+				var entity = sesTen1.Query<Entity>().WithOptions(x => x.SetCacheable(true)).Where(e => e.Id == _id).SingleOrDefault();
+				sesTen1.Disconnect();
+				deserializedSession = SpoofSerialization(sesTen1);
+			}
+
+			Sfi.Statistics.Clear();
+			using (deserializedSession)
+			{
+				deserializedSession.Reconnect();
+
+				//Expect session cache hit
+				var entity = deserializedSession.Get<Entity>(_id);
+				if (IsSqlServerDialect)
+					ValidateSqlServerConnectionAppName(deserializedSession, "tenant1");
+				deserializedSession.Clear();
+
+				//Expect second level cache hit
+				deserializedSession.Get<Entity>(_id);
+				Assert.That(GetTenantId(deserializedSession), Is.EqualTo("tenant1"));
+			}
+
+			Assert.That(Sfi.Statistics.PrepareStatementCount, Is.EqualTo(0));
+			Assert.That(Sfi.Statistics.SecondLevelCacheHitCount, Is.EqualTo(1));
+		}
+
+		private static string GetTenantId(ISession deserializedSession)
+		{
+			return deserializedSession.GetSessionImplementation().GetTenantIdentifier();
+		}
```

Comment:
```suggestion
		private static string GetTenantId(ISession session)
		{
			return session.GetSessionImplementation().GetTenantIdentifier();
		}
```

This is not specific to deserialized sessions, isn't it?